### PR TITLE
Fixed #29988 -- Updated coding style allow f-strings.

### PR DESCRIPTION
**Pull request by carltongibson**
This has come up again on #12646. This is a suggestion based on the previous discussion on #12650, and follow up with @felixxm. 

I will post to the mailing list — all objections there 😀
**Update**: [Discussion here](https://groups.google.com/d/topic/django-developers/XNhpOHam0zE/discussion)

### CODE REVIEW

**Comment by carltongibson** [comment]
Added https://github.com/django/django/pull/13214/commits/893ebb5de252e7d1e3a6e43e041e82ff28f747fd just to say that `format()` is not preferred. (This is just a possibility based on the list discussion.)

**Comment by jdufresne** [comment]
Thoughts on making it explicit that nested quotes should be avoided? IMO, that is an f-string edge case that is slightly less readable. For example, don't allow:

```py
v = f"{foo['bar']}"
```

**Comment by pope1ni** [comment]
Hi Carlton,

So I think it would be nice to get this finalised and landed given that there are a number of pull requests turning up making use of f-strings and we haven't quite nailed this down yet. It doesn't feel like there is much left to go.

To summarise the changes so far:

- Any of %-formatting, ``format()`` or f-strings are allowed with the goal of maximising readability.
- No f-strings for translatable strings and error or log messages.
- ``format()`` is more verbose so generally discouraged.
- Don't waste time doing unrelated refactoring of existing code to adjust the method used.

I agree with all these points. I have the following thoughts for consideration;

- Performance. ``.format()`` can be terrible compared to f-strings, it's not just about the verbosity. It is also often worse than %-formatting. Do we want to mention anything regarding change for performance reasons as acceptable - on a case-by-case basis, not in bulk - if accompanied by evidence? Or just leave it implicit? Perhaps at least worth mentioning that ``.format()`` isn't great.
- My [comment](https://github.com/django/django/pull/13214#pullrequestreview-452272966) above about adding appropriate links through to the Python documentation. This will be useful for those who find it difficult to get to the right sections (me) and those who are less au fait with the different options available.
- Please see my [comment](https://groups.google.com/g/django-developers/c/XNhpOHam0zE/m/x5TrIXwJBgAJ) on the mailing list about correcting the documentation regarding f-strings and xgettext. I think that we should be a bit clearer.
- Do we want to explicitly mention that the oft-forgotten [template strings](https://docs.python.org/3/library/string.html#template-strings) are undesired?
- Finally, perhaps we should add a little code block of what constitutes readable. There have been various suggestions, e.g. @jdufresne has proposed [no subscripting if quotes are required](https://github.com/django/django/pull/13214#issuecomment-661826671), and @charettes has proposed [excluding multiplication and function/method calls](https://github.com/django/django/pull/13214#issuecomment-661826671). I personally feel that method calls are fine within reason, but function calls are not. Perhaps something like the following as a starting point:
  ```python
  # Good:
  f'Hello, {request.user.username}!'
  f'Hello, {user.get_full_name()}!'
  f'Hello, {user.username.upper()}!'
  f'Hello, {users[0].username}!'
  key = 'manager'; f'Hello, {users[key].username}!'
  age = user.age * 365.25; f'Hello {user.get_full_name()}. You are {age:,d} days old!'

  # Bad:
  f"{request['Content-Type']}"  # Quotes in subscript make string harder to read.
  f'{request.user}'  # Just use str(request.user) instead if no surrounding text.
  f'{request.user!r}'  # Just use repr(request.user) instead if no surrounding text.
  f'{repr(request.user)}'  # Just use repr(request.user) instead if no surrounding text.
  f'{get_username()}'  # Function return value could be returned directly.
  f'Hello, {get_username()}'  # Function return value should be assigned to a local variable.
  f'You are {user.age * 365.25} days old.'  # Complex expressions should be assigned to a local variable.
  f'Hello, {self.request.user.username}!'  # Too many chained attribute lookups. Two is a reasonable number.
  f'Hello, {username.lower().strip()}!'  # Too many chained method calls.
  ```

Thanks for taking this on. :beers:

**Comment by carltongibson** [comment]
Hey @pope1ni. Thanks for looking at this. Yes, we should resolve...

* "Readable" is the difficult one. I'll have a think about, but also would take, a sentence say ≈"What's readable will be left to the Merger's discretion. Here's a couple of examples that would be excluded..." — I can't think we can do better than that (nor should we try IMO). (But... 🍿)
* I wouldn't mention performance or template strings. (I think both are a distraction.) 
* Happy to add links, yes. 
* Similar with xgettext comment. No problem. 

I'll pick this up. Ta.

**Comment by carltongibson** [comment]
Hey @pope1ni. I've updated based on your comments. Given other opinions, I've erred slightly more to the side of caution than you suggested, but I hope that won't upset you. 

Ref https://github.com/django/django/commit/c3437f734d03d93f798151f712064394652cabed I looked at adjusting to "Python f-strings_ **that require translation** … are not yet supported by `xgettext`" — but this seems redundant, translation being the whole point of `xgettext`. If you can't use an f-string with a substitution then I think "not yet supported by `xgettext`" is fair, and that making the exact issue explicit is likely more noise that it's worth. The bottom line is you can't use it right? — Happy to look at a suggestion there if you've got one (small/simple).

**Comment by carltongibson** [comment]
Does help if you actually push the changes. ☕️

**Comment by carltongibson** [comment]
> In British English the normal spelling in general contexts is judgement. However, the spelling judgment is conventional in legal contexts, and in North American English.

File: docs/internals/contributing/writing-code/coding-style.txt [hunk 1]
```diff
@@ -52,6 +52,41 @@ Python style
...
+
+    # Disallowed
+    f'hello {get_user()}'
+    f'you are {user.age * 365.25} days old'
+
+    # Allowed with local variable assignment
+    user = get_user()
+    f'hello {user}'
+    user_days_old = user.age * 365.25
+    'you are {user_days_old} days old'
```
Comment (pope1ni) [inline docs/internals/contributing/writing-code/coding-style.txt:76]:
```suggestion
    f'you are {user_days_old} days old'
```
:smirk:

File: docs/internals/contributing/writing-code/coding-style.txt [hunk 2]
```diff
@@ -89,6 +124,9 @@ Python style
         """
         ...
 
+
+
+
```
Comment (pope1ni) [inline docs/internals/contributing/writing-code/coding-style.txt:129]:
```suggestion
```

(Huh. The suggestion only shows removal of one line, not three.)

**Review by pope1ni — APPROVED** [review]
Thanks Carlton. This is great! :+1:

**Comment by pope1ni** [comment]
> ..., but I hope that won't upset you.

:cry: :stuck_out_tongue: 

> In British English the normal spelling in general contexts is judgement. However, the spelling judgment is conventional in legal contexts, and in North American English.

:grimacing: :see_no_evil: At least that didn't refer to "English" (US) and "International English" (British) like so many software installers of the 90's!

File: docs/internals/contributing/writing-code/coding-style.txt (hunk 2 above)
Comment (carltongibson) [reply docs/internals/contributing/writing-code/coding-style.txt]: I was gonna put links there. Thanks!

**Review by smithdc1 — APPROVED** [review]

File: docs/internals/contributing/writing-code/coding-style.txt [hunk 3]
```diff
@@ -52,6 +52,41 @@ Python style
   single quote. Don't waste time doing unrelated refactoring of existing code
   to conform to this style.
 
+* String variable interpolation may use `%-formatting`__, `f-strings`__, or
+  :py:meth:`str.format` as appropriate, with the goal of maximizing code
+  readability.
```
Comment (felixxm) [inline docs/internals/contributing/writing-code/coding-style.txt:57]: You can use `:py:ref:`:
```suggestion
* String variable interpolation may use
  :py:ref:`%-formatting <old-string-formatting>`,
  :py:ref:`f-strings <f-strings>`, or :py:meth:`str.format` as appropriate,
  with the goal of maximizing code readability.
```

File: docs/internals/contributing/writing-code/coding-style.txt [hunk 4]
```diff
@@ -52,6 +52,41 @@ Python style
...
+    user = get_user()
+    f'hello {user}'
+    user_days_old = user.age * 365.25
+    f'you are {user_days_old} days old'
+
+  f-strings should not be used for any string that may require translation,
+  including error and logging messages.
+
+  In general ``format()`` is more verbose, so the other formatting methods are
+  preferred.
```
Comment (felixxm) [inline docs/internals/contributing/writing-code/coding-style.txt:82]: I would chop blank line.
```suggestion
  f-strings should not be used for any string that may require translation,
  including error and logging messages. In general ``format()`` is more
  verbose, so the other formatting methods are preferred.
```

**Review by felixxm — APPROVED** [review]
Thanks Carlton :+1:

**Review by claudep — APPROVED** [review]
Sounds good!

File: docs/internals/contributing/writing-code/coding-style.txt (hunk 3 above)
Comment (carltongibson) [reply docs/internals/contributing/writing-code/coding-style.txt]: Super. If only Sphinx could make coffee. 😀

File: docs/internals/contributing/writing-code/coding-style.txt (hunk 4 above)
Comment (carltongibson) [reply docs/internals/contributing/writing-code/coding-style.txt]: Done. Thanks.

**Comment by carltongibson** [comment]
Super, thanks all. Just waiting for CI now.

**Pull request merged**